### ISSUE TITLE
Disallow conditions that set static values within dynamic styles

### ISSUE THREAD

**Issue by nmn** [facebook/stylex#727]

**Comment by necolas** [comment]
Could this become an automatic compiler optimization for functional style rules? It would be aligned with where UX/DX perf optimizations are heading with React. The function syntax is quite nice, even for conditional styles that might not be dynamic.

If the function rule is used in the same file where it's defined...

```jsx
const styles = stylex.create({
  root: (isHighlighted) => ({
    color: isHighlighted ? 'grey' : 'black',
  })
});

function Foo() {
  return <html.div style={styles.root(isHighlighted))} />
}
```

It would ideally be transformed to the same output as the "conditional" syntax:

```jsx
const styles = {
  root: {
    $$css: true,
    color: 'xdjwem',
  },
  root_isHighlighted: {
    $$css: true,
    color: 'xpqemf',
  }
});

function Foo() {
  return <html.div style={[ styles.root, isHighlighted && styles.root_isHighlighted  ]} />
}
```

The function rule could be defined in one file and used in another. 

```jsx
import { menuStyles } from './styles'

function Menu() {
  return <html.div style={menuStyles.root(isExpanded)} />
}
```

And potentially "memoized" to class names in many cases.

```jsx
const memoizedStyles = {
  root: {
    $$css: true,
    color: 'xdjwem',
  },
  root_isExpanded: {
    $$css: true,
    color: 'xpqemf',
  }
});
  
export const menuStyles = {
  root: (isExpanded) => isExpanded ? memoizedStyles.root_isExpanded : memoizedStyles.root
});
```

Using the function rules this way is currently discouraged, but if the compiler could make it more efficient, it could be quite an ergonomic API for libraries using StyleX.

**Comment by nmn** [comment]
> Could this become an automatic compiler optimization for functional style rules?

To some extent yes. I'm currently thinking through the patterns that could be detected and optimised. 

Some half-formed thoughts:
- The last example you shared, where the function is preserved with "memoized" objects that are hoisted out is fairly achievable
- It would work best when *all* styles within the function are conditional. A mix of conditional and "truly dynamic" styles would be problematic.
- We would need to detect the *number* of possible conditions used within the function, as too many conditions within the same function optimised would lead to a combinatorial explosion of code.
  - When optimising `stylex.props` with statically known styles, we bail out after 4 conditions.

**Comment by necolas** [comment]
Breaking conditional declarations out into individual rules and would avoid combinatorial complexity. It's just like what you'd do by hand if required to manually convert it to the existing "conditional styles" pattern

**Comment by nmn** [comment]
Yeah, that's a great idea and it's something that should be doable without a huge refactor. Let me create a new issue for this.
